# 5 Modeling<a id='5_Modeling'></a>

## 5.3 Imports<a id='5.3_Imports'></a>

In [1]:
import pandas as pd
import numpy as np
import os
import pickle
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import __version__ as sklearn_version
from sklearn.model_selection import cross_validate

In [2]:
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from sklearn.pipeline import make_pipeline
from sklearn.feature_selection import SelectKBest, f_regression
import datetime

In [3]:
#return the first 5 and last 5 rows of this dataframe
def p(df_):
    if df_.shape[0] > 6:
        print(df_.shape)
        return pd.concat([df_.head(), df_.tail()])
    else:
        return df_

def rcp(file_, pd_=None, ic_=None):
    if (pd_ == None) and (ic_ == None):
        return pd.read_csv(os.path.join('..', 'processed_data', file_))
    elif (pd_ != None) and (ic_ == None):
        return pd.read_csv(os.path.join('..', 'processed_data', file_), parse_dates=pd_)
    elif (pd_ != None) and (ic_ != None):
        return pd.read_csv(os.path.join('..', 'processed_data', file_), parse_dates=pd_, index_col = ic_)
    else:
        return pd.read_csv(os.path.join('..', 'processed_data', file_), index_col = ic_)
    
def rcr(file_, pd_=None):
    if pd_ == None:
        return pd.read_csv(os.path.join('..', 'raw_data', file_))
    else:
        return pd.read_csv(os.path.join('..', 'raw_data', file_), parse_dates=pd_)
    
#sort dataframe by column
def s(df_, column_):
    return df_.sort_values(column_)

#reset index and sort dataframe by column
def sr(df_, column_, ascending_=True):
    df_ = df_.sort_values(column_, ascending=ascending_)
    return df_.reset_index(drop=True)

#print length of list
def pl(list_):
    print(len(list_))
    return list_

#print length of dictionary
def pdc(dict_):
    print(len(dict_))
    return dict_

def MAPE(Y_actual,Y_Predicted):
    mape = np.mean(np.abs((Y_actual - Y_Predicted)/Y_actual))*100
    return mape

## 5.4 Load Model<a id='5.4_Load_Model'></a>

In [4]:
# This isn't exactly production-grade, but a quick check for development
# These checks can save some head-scratching in development when moving from
# one python environment to another, for example
expected_model_version = '1.0'
model_path = os.path.join('..','models', '173C_linear_regression_select_k_best_model_trial_1.pkl')
if os.path.exists(model_path):
    with open(model_path, 'rb') as f:
        lr_selectkbest_model = pickle.load(f)
    if lr_selectkbest_model.version != expected_model_version:
        print("Expected model version doesn't match version loaded")
    if lr_selectkbest_model.sklearn_version != sklearn_version:
        print("Warning: model created under different sklearn version")
else:
    print("Expected model not found")

In [5]:
lr_selectkbest_model

Pipeline(steps=[('selectkbest',
                 SelectKBest(k=89,
                             score_func=<function f_regression at 0x7ffbbb57d820>)),
                ('linearregression', LinearRegression())])

In [6]:
# This isn't exactly production-grade, but a quick check for development
# These checks can save some head-scratching in development when moving from
# one python environment to another, for example
expected_model_version = '1.0'
model_path = os.path.join('..','models', '173C_model_trial_1.pkl')
if os.path.exists(model_path):
    with open(model_path, 'rb') as f:
        rf_model = pickle.load(f)
    if rf_model.version != expected_model_version:
        print("Expected model version doesn't match version loaded")
    if rf_model.sklearn_version != sklearn_version:
        print("Warning: model created under different sklearn version")
else:
    print("Expected model not found")

/opt/anaconda3/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeRegressor from version 0.22 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.
  warnings.warn(
/opt/anaconda3/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator RandomForestRegressor from version 0.22 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.
  warnings.warn(
/opt/anaconda3/lib/python3.8/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator Pipeline from version 0.22 when using version 0.23.2. This might lead to breaking code or invalid results. Use at your own risk.
  warnings.warn(


In [7]:
rf_model

Pipeline(steps=[('randomforestregressor',
                 RandomForestRegressor(criterion='mae', max_depth=10,
                                       min_samples_split=100, n_estimators=200,
                                       random_state=47))])

## 5.5 Load Data<a id='5.5_Load_Data'></a>

In [8]:
df_train = rcp('173C_train_split_time_series_merge_173A_and_173B.csv', None)
p(df_train)

(9239, 101)


,travel_distance_diff,latitude_diff,longitude_diff,team_id_b_1610612759,games_back_to_back_percentage_game_window999_diff,city_proper_metro_Area_b_San Antonio,future_Away_game_percent_day_window999_diff,future_home_game_percent_day_window999_diff,away_game_percent_day_window999_diff,games_back_to_back_count_game_window999_diff,...,NET_RATING_cma12_diff,pf_cmax999_diff,lc_sum_NET_RATING_cmax999_diff,lc_max_AST_TOV_cmax12_diff,lc_sum_AST_PCT_cmax5_diff,fg3_pct_cma999_diff,ft_pct_cmax12_diff,lc_max_MIN_cmax5_diff,E_NET_RATING_cma5_diff,E_NET_RATING_cmax12_diff
0,-1.045984,2.249661,0.309284,0,-0.016676,0,0.009482,-0.009482,-6.465582,-0.052324,...,-0.026252,0.110128,-0.049559,-0.088860,-0.094422,-0.008676,0.001855,-0.025856,-0.017154,-0.031080
1,0.835137,-1.609206,0.383027,0,-0.016676,0,0.009482,-0.009482,6.416819,-0.052324,...,-0.026252,0.110128,-0.049559,-0.088860,-0.094422,-0.008676,0.001855,-0.025856,-0.017154,-0.031080
2,1.138839,-0.562003,1.001028,0,-0.016676,0,0.009482,-0.009482,6.416819,-0.052324,...,-0.026252,0.110128,-0.049559,-0.088860,-0.094422,-0.008676,0.001855,-0.025856,-0.017154,-0.031080
3,-1.256725,0.877570,1.270549,0,-0.016676,0,0.009482,-0.009482,-6.465582,-0.052324,...,-0.026252,0.110128,-0.049559,-0.088860,-0.094422,-0.008676,0.001855,-0.025856,-0.017154,-0.031080
4,0.848535,-1.892884,-0.403891,0,6.083854,0,-0.030497,0.030497,6.416819,0.414362,...,-0.879667,4.661921,-0.319484,1.247488,1.691365,8.286999,8.007766,3.521210,-0.677803,-0.598734
9234,1.950338,0.908407,-2.448327,0,1.781375,0,0.163685,-0.163685,-0.176936,1.347733,...,-0.234987,-0.973632,-1.004135,-2.694313,0.931848,0.273061,0.163264,-0.448480,0.130420,0.013153
9235,-0.159237,-0.346823,-0.114386,0,0.464945,0,0.163685,-0.163685,-0.838007,0.414362,...,-0.967409,-0.540128,-1.002372,0.228034,1.378510,0.025707,0.196310,-0.184705,-0.567981,0.543946
9236,0.668894,0.365307,0.609586,0,-1.933986,0,0.059225,-0.059225,0.113645,-1.452381,...,0.552851,0.543632,-1.278631,0.481034,-0.087806,-1.030682,-0.349282,1.555168,-0.463307,-0.900991
9237,-1.178771,-1.458661,-1.226125,0,-0.016676,0,-0.395302,0.395302,0.992650,-0.052324,...,1.152273,0.977137,-0.462101,5.274412,-1.226123,0.139954,-0.431214,0.707843,1.522073,0.507086
9238,-1.593147,1.080693,-1.772901,0,1.203430,0,0.009482,-0.009482,-0.346441,0.881048,...,1.435821,-0.540128,-0.305763,0.053790,-0.145745,1.330639,0.001855,0.458857,0.813377,1.745603


In [9]:
df_test = rcp('173C_test_split_time_series_merge_173A_and_173B.csv', None)
p(df_test)

(1027, 101)


,travel_distance_diff,latitude_diff,longitude_diff,team_id_b_1610612759,games_back_to_back_percentage_game_window999_diff,city_proper_metro_Area_b_San Antonio,future_Away_game_percent_day_window999_diff,future_home_game_percent_day_window999_diff,away_game_percent_day_window999_diff,games_back_to_back_count_game_window999_diff,...,NET_RATING_cma12_diff,pf_cmax999_diff,lc_sum_NET_RATING_cmax999_diff,lc_max_AST_TOV_cmax12_diff,lc_sum_AST_PCT_cmax5_diff,fg3_pct_cma999_diff,ft_pct_cmax12_diff,lc_max_MIN_cmax5_diff,E_NET_RATING_cma5_diff,E_NET_RATING_cmax12_diff
0,1.028423,0.322683,1.055866,0,-0.423378,0,-0.094902,0.094902,0.619739,-0.519009,...,-0.323654,0.543632,-0.623327,0.183842,0.547129,-0.093945,-0.615885,0.401065,-1.033010,-1.505506
1,-0.989872,2.184396,-0.524758,0,1.374673,0,-0.239616,0.239616,0.521802,0.881048,...,-1.365485,-0.106624,-1.258891,-0.790590,2.544037,-0.582190,0.316977,1.269063,-1.523778,0.278550
2,0.792883,0.757034,0.660032,0,0.625485,0,-0.091714,0.091714,0.653640,0.414362,...,0.285004,0.326880,1.028775,-0.317461,0.295721,0.448738,-1.731027,-0.381070,0.322609,-0.067941
3,-1.063182,1.568692,0.796300,0,-1.091531,0,-0.355300,0.355300,0.742428,-0.985695,...,1.161510,-0.756880,0.248091,1.364641,0.100918,0.879725,0.404729,0.836795,0.753318,0.853576
4,-0.509033,0.020163,0.452923,0,-0.921539,0,-0.149777,0.149777,0.098895,-0.519009,...,0.792989,-0.973632,0.899901,0.785766,-1.158458,-0.596616,-1.233626,1.533060,1.247518,-0.001591
1022,1.359732,1.106411,-1.506188,0,-0.016676,0,0.009482,-0.009482,0.040681,-0.052324,...,0.854871,-0.973632,0.731619,-1.495076,0.110108,0.406319,0.038192,-0.733868,1.695387,1.480207
1023,1.789289,0.507847,1.893345,0,-0.154391,0,0.009482,-0.009482,0.071918,-0.519009,...,-0.452036,0.543632,1.359431,0.365893,0.047835,-0.323504,-1.048304,-0.077045,-0.640052,-0.849386
1024,1.789289,0.507847,1.893345,0,-0.152875,0,0.009482,-0.009482,0.134429,-0.519009,...,-0.542549,0.543632,1.359431,0.365893,0.047835,-0.339029,-1.110078,-0.077045,-0.730998,-0.849386
1025,-1.790571,0.507847,1.893345,0,-0.151392,0,0.009482,-0.009482,0.070036,-0.519009,...,-0.925847,0.543632,1.359431,0.392504,0.590597,-0.352941,-1.110078,1.181174,-1.566677,-0.849386
1026,-1.790571,0.507847,1.893345,0,-0.149940,0,0.009482,-0.009482,0.006887,-0.519009,...,-1.439374,0.543632,1.359431,0.392504,0.590597,-0.360802,-1.110078,1.195694,-1.915019,-0.849386


In [10]:
X_train = df_train.drop(columns = 'spread_a')
p(X_train)

(9239, 100)


,travel_distance_diff,latitude_diff,longitude_diff,team_id_b_1610612759,games_back_to_back_percentage_game_window999_diff,city_proper_metro_Area_b_San Antonio,future_Away_game_percent_day_window999_diff,future_home_game_percent_day_window999_diff,away_game_percent_day_window999_diff,games_back_to_back_count_game_window999_diff,...,NET_RATING_cma12_diff,pf_cmax999_diff,lc_sum_NET_RATING_cmax999_diff,lc_max_AST_TOV_cmax12_diff,lc_sum_AST_PCT_cmax5_diff,fg3_pct_cma999_diff,ft_pct_cmax12_diff,lc_max_MIN_cmax5_diff,E_NET_RATING_cma5_diff,E_NET_RATING_cmax12_diff
0,-1.045984,2.249661,0.309284,0,-0.016676,0,0.009482,-0.009482,-6.465582,-0.052324,...,-0.026252,0.110128,-0.049559,-0.088860,-0.094422,-0.008676,0.001855,-0.025856,-0.017154,-0.031080
1,0.835137,-1.609206,0.383027,0,-0.016676,0,0.009482,-0.009482,6.416819,-0.052324,...,-0.026252,0.110128,-0.049559,-0.088860,-0.094422,-0.008676,0.001855,-0.025856,-0.017154,-0.031080
2,1.138839,-0.562003,1.001028,0,-0.016676,0,0.009482,-0.009482,6.416819,-0.052324,...,-0.026252,0.110128,-0.049559,-0.088860,-0.094422,-0.008676,0.001855,-0.025856,-0.017154,-0.031080
3,-1.256725,0.877570,1.270549,0,-0.016676,0,0.009482,-0.009482,-6.465582,-0.052324,...,-0.026252,0.110128,-0.049559,-0.088860,-0.094422,-0.008676,0.001855,-0.025856,-0.017154,-0.031080
4,0.848535,-1.892884,-0.403891,0,6.083854,0,-0.030497,0.030497,6.416819,0.414362,...,-0.879667,4.661921,-0.319484,1.247488,1.691365,8.286999,8.007766,3.521210,-0.677803,-0.598734
9234,1.950338,0.908407,-2.448327,0,1.781375,0,0.163685,-0.163685,-0.176936,1.347733,...,-0.234987,-0.973632,-1.004135,-2.694313,0.931848,0.273061,0.163264,-0.448480,0.130420,0.013153
9235,-0.159237,-0.346823,-0.114386,0,0.464945,0,0.163685,-0.163685,-0.838007,0.414362,...,-0.967409,-0.540128,-1.002372,0.228034,1.378510,0.025707,0.196310,-0.184705,-0.567981,0.543946
9236,0.668894,0.365307,0.609586,0,-1.933986,0,0.059225,-0.059225,0.113645,-1.452381,...,0.552851,0.543632,-1.278631,0.481034,-0.087806,-1.030682,-0.349282,1.555168,-0.463307,-0.900991
9237,-1.178771,-1.458661,-1.226125,0,-0.016676,0,-0.395302,0.395302,0.992650,-0.052324,...,1.152273,0.977137,-0.462101,5.274412,-1.226123,0.139954,-0.431214,0.707843,1.522073,0.507086
9238,-1.593147,1.080693,-1.772901,0,1.203430,0,0.009482,-0.009482,-0.346441,0.881048,...,1.435821,-0.540128,-0.305763,0.053790,-0.145745,1.330639,0.001855,0.458857,0.813377,1.745603


In [11]:
y_train = df_train.spread_a
p(y_train)

(9239,)


0       -8
1       14
2        2
3        1
4      -10
9234    -3
9235   -16
9236   -10
9237   -15
9238   -15
Name: spread_a, dtype: int64

In [12]:
X_test = df_test.drop(columns = 'spread_a')
p(X_test)

(1027, 100)


,travel_distance_diff,latitude_diff,longitude_diff,team_id_b_1610612759,games_back_to_back_percentage_game_window999_diff,city_proper_metro_Area_b_San Antonio,future_Away_game_percent_day_window999_diff,future_home_game_percent_day_window999_diff,away_game_percent_day_window999_diff,games_back_to_back_count_game_window999_diff,...,NET_RATING_cma12_diff,pf_cmax999_diff,lc_sum_NET_RATING_cmax999_diff,lc_max_AST_TOV_cmax12_diff,lc_sum_AST_PCT_cmax5_diff,fg3_pct_cma999_diff,ft_pct_cmax12_diff,lc_max_MIN_cmax5_diff,E_NET_RATING_cma5_diff,E_NET_RATING_cmax12_diff
0,1.028423,0.322683,1.055866,0,-0.423378,0,-0.094902,0.094902,0.619739,-0.519009,...,-0.323654,0.543632,-0.623327,0.183842,0.547129,-0.093945,-0.615885,0.401065,-1.033010,-1.505506
1,-0.989872,2.184396,-0.524758,0,1.374673,0,-0.239616,0.239616,0.521802,0.881048,...,-1.365485,-0.106624,-1.258891,-0.790590,2.544037,-0.582190,0.316977,1.269063,-1.523778,0.278550
2,0.792883,0.757034,0.660032,0,0.625485,0,-0.091714,0.091714,0.653640,0.414362,...,0.285004,0.326880,1.028775,-0.317461,0.295721,0.448738,-1.731027,-0.381070,0.322609,-0.067941
3,-1.063182,1.568692,0.796300,0,-1.091531,0,-0.355300,0.355300,0.742428,-0.985695,...,1.161510,-0.756880,0.248091,1.364641,0.100918,0.879725,0.404729,0.836795,0.753318,0.853576
4,-0.509033,0.020163,0.452923,0,-0.921539,0,-0.149777,0.149777,0.098895,-0.519009,...,0.792989,-0.973632,0.899901,0.785766,-1.158458,-0.596616,-1.233626,1.533060,1.247518,-0.001591
1022,1.359732,1.106411,-1.506188,0,-0.016676,0,0.009482,-0.009482,0.040681,-0.052324,...,0.854871,-0.973632,0.731619,-1.495076,0.110108,0.406319,0.038192,-0.733868,1.695387,1.480207
1023,1.789289,0.507847,1.893345,0,-0.154391,0,0.009482,-0.009482,0.071918,-0.519009,...,-0.452036,0.543632,1.359431,0.365893,0.047835,-0.323504,-1.048304,-0.077045,-0.640052,-0.849386
1024,1.789289,0.507847,1.893345,0,-0.152875,0,0.009482,-0.009482,0.134429,-0.519009,...,-0.542549,0.543632,1.359431,0.365893,0.047835,-0.339029,-1.110078,-0.077045,-0.730998,-0.849386
1025,-1.790571,0.507847,1.893345,0,-0.151392,0,0.009482,-0.009482,0.070036,-0.519009,...,-0.925847,0.543632,1.359431,0.392504,0.590597,-0.352941,-1.110078,1.181174,-1.566677,-0.849386
1026,-1.790571,0.507847,1.893345,0,-0.149940,0,0.009482,-0.009482,0.006887,-0.519009,...,-1.439374,0.543632,1.359431,0.392504,0.590597,-0.360802,-1.110078,1.195694,-1.915019,-0.849386


In [13]:
y_test = df_test.spread_a
p(y_test)

(1027,)


0       13
1        7
2      -10
3      -11
4       10
1022    -9
1023    10
1024    19
1025     8
1026    23
Name: spread_a, dtype: int64

In [14]:
y_test_pred_rf = rf_model.predict(X_test)
y_test_pred_rf

array([4.955 , 6.3175, 5.63  , ..., 9.0675, 3.075 , 3.4425])

In [15]:
mean_absolute_error(y_test, y_test_pred_rf)

9.683213242453748

In [16]:
MAPE(y_test, y_test_pred_rf)

112.86574956201913

In [19]:
y_test_pred_lr_selectkbest = lr_selectkbest_model.predict(X_test)
y_test_pred_lr_selectkbest

array([4.24169134, 4.26717477, 6.7927086 , ..., 6.44600583, 1.5854303 ,
       3.22124366])

In [20]:
mean_absolute_error(y_test, y_test_pred_lr_selectkbest)

9.686634225919326

In [21]:
MAPE(y_test, y_test_pred_lr_selectkbest)

111.45737300662336

## 5.9 Concluding Thoughts<a id='5.9_Modeling_scenarios'></a>

When comparing the performance of the linear regression and random forest models, there is a potential preference for the random forest model because of the lower mean absolute error compared to the linear regression model at 9.683 and 9.686 respectively.